erts/emulator/nifs/common/prim_tty_nif.c

```diff
@@ -0,0 +1,1026 @@
+/*
+ * %CopyrightBegin%
+ *
+ * Copyright Ericsson 2015-2021. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * %CopyrightEnd%
+ */
+
+/*
+ * Purpose:  NIF library for interacting with the tty
+ *
+ */
+
+#define STATIC_ERLANG_NIF 1
+
+#ifndef WANT_NONBLOCKING
+#define WANT_NONBLOCKING
+#endif
+
+#include "config.h"
+#include "sys.h"
+#include "erl_nif.h"
+#include "erl_driver.h"
+
+#include <errno.h>
+#include <string.h>
+#include <ctype.h>
+#include <wchar.h>
+#include <stdio.h>
+#include <signal.h>
+#include <locale.h>
+#ifdef HAVE_TERMCAP
+  #include <termios.h>
+  #include <curses.h>
+  #include <term.h>
+#endif
+#ifndef __WIN32__
+  #include <unistd.h>
+  #include <sys/ioctl.h>
+#endif
+#ifdef HAVE_SYS_UIO_H
+  #include <sys/uio.h>
+#endif
+
+#if defined IOV_MAX
+#define MAXIOV IOV_MAX
+#elif defined UIO_MAXIOV
+#define MAXIOV UIO_MAXIOV
+#else
+#define MAXIOV 16
+#endif
+
+#if !defined(HAVE_SETLOCALE) || !defined(HAVE_NL_LANGINFO) || !defined(HAVE_LANGINFO_H)
+#define PRIMITIVE_UTF8_CHECK 1
+#else
+#include <langinfo.h>
+#endif
+
+#ifdef VALGRIND
+#  include <valgrind/memcheck.h>
+#endif
+
+#define DEF_HEIGHT 24
+#define DEF_WIDTH 80
+
+typedef struct {
+#ifdef __WIN32__
+    HANDLE ofd;
+    HANDLE ifd;
+    DWORD dwOriginalOutMode;
+    DWORD dwOriginalInMode;
+    DWORD dwOutMode;
+    DWORD dwInMode;
+#else
+    int ofd;       /* stdout */
+    int ifd;       /* stdin */
+#endif
+    ErlNifPid self;
+    ErlNifPid reader;
+    int tty;       /* if the tty is initialized */
+#ifdef THREADED_READER
+    ErlNifTid reader_tid;
+#endif
+#ifndef __WIN32__
+    int signal[2]; /* Pipe used for signal (winch + cont) notifications */
+#endif
+#ifdef HAVE_TERMCAP
+    struct termios tty_smode;
+    struct termios tty_rmode;
+#endif
+} TTYResource;
+
+static ErlNifResourceType *tty_rt;
+
+/* The NIFs: */
+static ERL_NIF_TERM isatty_nif(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[]);
```

Comment:
Whats the naming convention here, when should you have _nif and when should you not